### PR TITLE
Add libstdc++-10 related mappings

### PR DESCRIPTION
**Pull request by saki7**
Checked for my application using C++20 flags, with Clang 13.

---

__Issue described below has been moved to: #884__

~~Although most of the diff is trivial, I still get some error in my code:~~

> ~~#include <filesystem>~~
>
> ~~void clean_file(const std::filesystem::path& path)~~
> ~~{~~
>      ~~remove(path); // ADL~~
> ~~}~~

~~`std::filesystem::path` and `std::filesystem::remove` both sit in the same namespace, `std::filesystem`. So calling the `remove` function without namespace qualification should trigger ADL. My code correctly compiles without any compiler warning on my actual build.~~

~~However IWYU suggests including `<stdio.h>` for the function `remove`. It is obvious that [`remove(const char*)`](https://en.cppreference.com/w/c/io/remove) (in global namespace) should not be found even with ADL; the more matching signature is [`remove(const std::filesystem::path&)`](https://en.cppreference.com/w/cpp/filesystem/remove) (in namespace std::filesystem). In IWYU, where does this strange suggestion come from?~~

~~The ADL issue seem unrelated to this specific PR, so I omitted the workaround inside mapping update. I guess there exists some nasty pitfalls for ADL detection in IWYU. Maybe we could discuss that in another issue?~~

### CODE REVIEW

**Comment by kimgr** [comment]
Thanks for the updated mappings!

Unfortunately we carry them in two places -- in gcc.stl.headers.imp and inline inside `iwyu_include_picker.cc`. Could you modify both?

Re the dual `remove`, I wonder if that's somehow related to:
https://github.com/include-what-you-use/include-what-you-use/issues/636

Let's open a separate issue for that, and maybe you can investigate the AST with `-Xiwyu -v7` and see if it's possible to identify where IWYU is going wrong?

**Comment by saki7** [comment]
@kimgr I applied updates to both `gcc.stl.headers.imp` and `iwyu_include_picker.cc`.

Most of the standard headers should be backward-compatible, so I assume that hard-coding C++20 headers in IWYU does not lead to real world issues. If you are happy with this decision, please go ahead and merge this PR.

For now I've chosen C++20 because it's the latest published standard. I used the `c++20` branch from `cplusplus/draft` repo, to generate the list. If we are going to be pedantic then we expect IWYU to automatically detect user's C++ language version flags at the IWYU invocation, so it switches the header resolution dynamically. But that's more like a new feature request, so I'm just sharing my idea here.

p.s. I will split the ADL bug into different issue.

File: gcc.stl.headers.imp [hunk 1]
```diff
@@ -157,15 +168,16 @@
   { include: ["<tr1/unordered_map.h>", private, "<tr1/unordered_map>", public ] },
   { include: ["<tr1/unordered_set.h>", private, "<tr1/unordered_set>", public ] },
   { include: ["<tr2/dynamic_bitset.tcc>", private, "<tr2/dynamic_bitset>", public ] },
-  # ( cd /usr/include/x86_64-linux-gnu/c++/8 && grep -r headername | perl -nle 'm/^([^:]+).*@headername\{([^,]*)\}/ && print qq@  { include: ["<$1>", private, "<$2>", public ] },@' | sort -u )
+  # ( cd /usr/include/x86_64-linux-gnu/c++/10 && grep -r headername | perl -nle 'm/^([^:]+).*@headername\{([^,]*)\}/ && print qq@  { include: ["<$1>", private, "<$2>", public ] },@' | sort -u )
   { include: ["<bits/basic_file.h>", private, "<ios>", public ] },
   { include: ["<bits/c++allocator.h>", private, "<memory>", public ] },
-  { include: ["<bits/c++config.h>", private, "<iosfwd>", public ] },
```
Comment (kimgr) [inline gcc.stl.headers.imp:163]: This one is a little scary -- I wonder if we should map this to both `<iosfwd>` _and_ `<version>` (I'm guessing `<version>` is new in libstdc++10).

File: gcc.stl.headers.imp [hunk 2]
```diff
@@ -300,4 +312,12 @@
   # The location of exception_defines.h varies by GCC version.  It should
   # never be included directly.
   { include: ["<exception_defines.h>", private, "<exception>", public ] },
+
+  # post libstdc++-10 stuffs which are not automatically caught by commands above
```
Comment (kimgr) [inline gcc.stl.headers.imp:316]: stuffs -> stuff
which are -> which is

File: iwyu_include_picker.cc [hunk 3]
```diff
@@ -473,9 +473,11 @@ const IncludeMapEntry stdlib_c_include_map[] = {
   // mapping.)  Here is how I identified the files to map:
   // $ for i in /usr/include/c++/4.4/c* ; do ls /usr/include/`basename $i | cut -b2-`.h /usr/lib/gcc/*/4.4/include/`basename $i | cut -b2-`.h 2>/dev/null ; done
   //
-  // These headers are defined in C++14 [headers]p3.  You can get them with
-  // $ sed -n '/begin{floattable}.*{tab:cpp.c.headers}/,/end{floattable}/p' lib-intro.tex | grep tcode | perl -nle 'm/tcode{<c(.*)>}/ && print qq@  { "<$1.h>", kPublic, "<c$1>", kPublic },@' | sort
-  // on https://github.com/cplusplus/draft/blob/master/source/lib-intro.tex
+  // These headers are defined in [headers.cpp.c].
+  // https://github.com/cplusplus/draft/blob/c+%2B20/source/lib-intro.tex
+  //
+  // $ wget https://raw.githubusercontent.com/cplusplus/draft/c%2B%2B20/source/lib-intro.tex
```
Comment (kimgr) [inline iwyu_include_picker.cc:479]: Nice! Could you fold this into a one-liner so the instruction doesn't require disk for lib-intro.tex? E.g.
```
$ curl https://raw.githubusercontent.com/cplusplus/draft/c%2B%2B20/source/lib-intro.tex | sed -n '/begin{multicolfloattable}.*{headers.cpp.c}/,/end{multicolfloattable}/p' ...
```

File: iwyu_include_picker.cc [hunk 4]
```diff
@@ -505,19 +507,31 @@ const IncludeMapEntry stdlib_c_include_map[] = {
 };
 
 const char* stdlib_cpp_public_headers[] = {
-  // These headers are defined in C++14 [headers]p2.  You can get them with
-  // $ sed -n '/begin{floattable}.*{tab:cpp.library.headers}/,/end{floattable}/p' lib-intro.tex | grep tcode | perl -nle 'm/tcode{(.*)}/ && print qq@  "$1",@' | sort
-  // on https://github.com/cplusplus/draft/blob/master/source/lib-intro.tex
+  // These headers are defined in [headers.cpp].
+  // https://github.com/cplusplus/draft/blob/c+%2B20/source/lib-intro.tex
+  //
+  // $ wget https://raw.githubusercontent.com/cplusplus/draft/c%2B%2B20/source/lib-intro.tex
```
Comment (kimgr) [inline iwyu_include_picker.cc:513]: Use curl to avoid storing lib-intro.text per above

File: iwyu_include_picker.cc [hunk 5]
```diff
@@ -721,10 +761,10 @@ const IncludeMapEntry libstdcpp_include_map[] = {
   { "<tr1/unordered_map.h>", kPrivate, "<tr1/unordered_map>", kPublic },
   { "<tr1/unordered_set.h>", kPrivate, "<tr1/unordered_set>", kPublic },
   { "<tr2/dynamic_bitset.tcc>", kPrivate, "<tr2/dynamic_bitset>", kPublic },
-  // cd /usr/include/x86_64-linux-gnu/c++/8 && grep -r headername | perl -nle 'm/^([^:]+).*@headername\{([^,]*)\}/ && print qq@  { "<$1>", kPrivate, "<$2>", kPublic },@' | sort -u
+  // cd /usr/include/x86_64-linux-gnu/c++/10 && grep -r headername | perl -nle 'm/^([^:]+).*@headername\{([^,]*)\}/ && print qq@  { "<$1>", kPrivate, "<$2>", kPublic },@' | sort -u
   { "<bits/basic_file.h>", kPrivate, "<ios>", kPublic },
   { "<bits/c++allocator.h>", kPrivate, "<memory>", kPublic },
-  { "<bits/c++config.h>", kPrivate, "<iosfwd>", kPublic },
```
Comment (kimgr) [inline iwyu_include_picker.cc:727]: Again with mapping to both. I think that would make sense.

**Review by kimgr — COMMENTED** [review]
Looks great. Some ideas for improvement inline.

**Comment by saki7** [comment]
Force-pushed. I believe all issues have been resolved.

**Comment by kimgr** [comment]
Thanks, LGTM!

**Pull request merged**